woocommerce/woocommerce-ios
# Product selector: Support multiple selections

Part of #6489 


### Description

As part of the Coupons project, we want to let merchants select products and variations for coupons. We're reusing the product selection screen from the Order Creation flow and updating it to support multiple selections and other customizations.

This PR focuses on handling multiple selections:
- `ProductRow`: 
  - Add new property to support multiple selections
  - Add a new image indicating the selection state of the row if multiple selections are enabled
  - Extra: add a processor to resize the product image in a background thread to avoid blocking the UI when the image is large.
  - Add a new enum for 3 selection states of the product row: selected, partial selected and not selected.
- `ProductSelector`: 
  - update configuration with 2 new configs: whether multiple selections are enabled and the format for the selection button.
  - Add an optional button at the bottom of the product list to complete multiple selections.
  - Send multiple-selection config to product rows and product variation selector.
- `ProductSelectorViewModel`: 
  - Add a new initializer to support multiple selections
  - Add support for multiple selections
- `ProductVariationSelector` and view model: update support for multiple selections.

Updates specific to Coupons features:
- Update product selector configs of AddEditCoupon and CouponRestrictions to support multiple selections.
- Update the AddEditCoupon view to display the total number of selected products or variations.

### Testing instructions

Testing Coupons flow:
- Make sure that your test store has coupons enabled and has set up at least one coupon in WP Admin > Marketing > Coupons.
- Open the app, make sure that coupon management is enabled in Menu > Settings > Experimental Features > Coupon Management.
- Navigate to Menu > Coupons and select a coupon.
- Tap the ellipsis button > Edit Coupon.
- On the Edit Coupon screen, if the coupon is currently applicable to all products, notice that the button reads "All Products". Otherwise, it should read "Edit Products (<number of selected products or variations>)".
- Tap that button, notice that a product selector view is presented. Try selecting/unselecting any product or variation to make sure that the selection works properly. 
- Tap the button at the bottom of the screen after selecting at least one item. Notice that the correct number of items is reflected on the product button of the Edit Coupon screen.
- Tap the Usage Restrictions row > Exclude Products. Notice that the multiple selections on the presented product selector view work properly too.

Please feel free to test the product selector on the New Order screen to make sure that single selection still works as before.


### Screenshots

https://user-images.githubusercontent.com/5533851/164448570-64580013-e5c3-4407-a00e-1082fd89e7b9.mp4


---
